### DISCUSSION TITLE
Units of totalVolRate  in WellControl and brine water salinity in FlashModel?

### DISCUSSION THREAD

**Discussion by wdmhouston** [GEOS-DEV/GEOS#1439] [Q&A] [answered]
Hi Experts,

    What are the units of totalVolRate  in WellControl and brine water salinity in FlashModel? m3/second and g/g?

Thanks and Regards,
wdm

**Comment by francoishamon** [comment, marked as answer]
In the block `WellControls`, when you select `control="totalVolRate"`, then the active constraint is set by `targetTotalRate="5e-8"`, where the value is in `m^3.s^-1`. Unless you select the flag `useSurfaceConditions=1`, this value is at reservoir conditions. I recommend setting `useSurfaceConditions=0` if you use the well to inject CO2. You can do something like:
```
   <WellControls
        name="wellControls"
        type="injector"
        control="totalVolRate"
        referenceElevation="-2000" 
        targetBHP="5e7"
        targetTotalRate="1e-1"
        injectionStream="{0.995, 0.005}"/>
```

In the `CO2BrineFluid` model, the water salinity is expressed in molality (molal concentration): moles of NaCl per kg of H2O. The model that we use is valid if the water salinity, m, is such that  m < 5 moles of NaCl per kg of H2O.

**Reply by wdmhouston** [reply]
Thanks! This helps!

**Reply by wdmhouston** [reply]
Francois,
  
    Here is my well perforation,
````
    <InternalWell
      name="wellInjector1"
      wellRegionName="wellRegion"
      wellControlsName="wellControls"
      meshName="mesh"
      polylineNodeCoords="{ { 5000, 2000, 50},
                            { 5000, 5000, 0} }"
      polylineSegmentConn="{ { 0, 1 } }"
      radius="0.09525"
      numElementsPerSegment="1">
      <Perforation
        name="injector1_perf1"
        distanceFromHead="17.3712"/>
    </InternalWell>
````
Does this mean this well's perforation is connected to well from  0 to 50-17.3712?

**Reply by wdmhouston** [reply]
never mind. I tested several cases and figured it out.

**Reply by francoishamon** [reply]
Ok, glad you found a solution. In the current setup, it means that the well trajectory is from { 5000, 2000, 50 } to { 5000, 2000, 0 }, with one perforation connecting the well with the reservoir cell located at { 5000, 2000, 32.6288 }

If you have an Eclipse `COMPDAT` keyword in your original deck that, say, looks like that (fake values below):
```
COMPDAT
--WELL        I     J    K1    K2 OP/SH  SATN       TRAN      WBDIA         KH       SKIN DFACT   DIR      PEQVR
-----------------------------------------------------------------------------------------------------------------
 'WELL_1'   22  16   41    41    OPEN  1*    0.15693041 0.2        7.61051403 20          1*    Z     3.76499936 /
 'WELL_1'   22  16   42    42    OPEN  1*    0.04412274 0.2        2.13978121 20          1*    Z     3.76500254 /
 'WELL_1'   22  16   43    43    OPEN  1*    1.33193269 0.2        64.5935559 20          1*    Z     3.76500867 /
```
here is what I suggest:

1. In your GRDECL mesh, you compute the reservoir cell center of cells (22,16,41), (22,16,42), (22,16,43)
2. You write the x, y, and z coordinates of the reservoir cell centers computed at step 1 at line 103 to 106 of the script attached.
3. You run the script that will generate an input for the well in the GEOSX format. 
4. You copy paste the script output into your xml. You make sure that the z coordinates in `polylineNodeCoords` are monotonically decreasing. You adjust the names to your naming choices. And this should work. 

If you want, you can also convert the `TRAN` column into the metric units, and give that value to GEOSX by doing:
```
<Perforation
        name="injector1_perf1"
        distanceFromHead="17.3712"
        transmissibility="1e-13"/>    <------- don't forget the conversion
```
Note that the script attached does not do that, and relies on GEOSX to compute an internal transmissibility.

If you want to use the script, you need to change the extension to .py, I cannot upload a .py file here
[find_perforations.txt](https://github.com/GEOSX/GEOSX/files/6580843/find_perforations.txt)
